# Aviation Accidents Analysis

You are part of a consulting firm that is tasked to do an analysis of commercial and passenger jet airline safety. The client (an airline/airplane insurer) is interested in knowing what types of aircraft (makes/models) exhibit low rates of total destruction and low likelihood of fatal or serious passenger injuries in the event of an accident. They are also interested in any general variables/conditions that might be at play. Your analysis will be based off of aviation accident data accumulated from the years 1948-2023. 

Our client is only interested in airplane makes/models that are professional builds and could potentially still be active. Assume a max lifetime of 40 years for a make/model retirement and make sure to filter your data accordingly (i.e. from 1983 onwards). They would also like separate recommendations for small aircraft vs. larger passenger models. **In addition, make sure that claims that you make are statistically robust and that you have enough samples when making comparisons between groups.**


In this summative assessment you will demonstrate your ability to:
- **Use Pandas to load, inspect, and clean the dataset appropriately.**
- **Transform relevant columns to create measures that address the problem at hand.**
- conduct EDA: visualization and statistical measures to systematically understand the structure of the data
- recommend a set of airplanes and makes conforming to the client's request and identify at least *two* factors contributing to airplane safety. You must provide supporting evidence (visuals, summary statistics, tables) for each claim you make.

### Make relevant library imports

In [178]:
#import the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

## Business Understanding
The goal of this analysis is to explore aviation accident data and identify patterns related to aircraft safety.

The client is interested in:
- Aircraft types (make/model) with lower rates of total destruction.
- Aircraft with a lower likelihood of fatal or serious injuries.
- General conditions that influence accident outcomes.

The analysis will focus on commercial and passenger aircraft and aim to provide insights that can support safer aircraft recommendations.

## Data Loading and Inspection

### Data Understanding
The dataset used in this analysis is sourced from aviation accident records maintained by the National Transportation Safety Board (NTSB).

It contains historical data on aircraft accidents and incidents, including details on aircraft characteristics, accident conditions, and injury outcomes.


### Load in data from the relevant directory and inspect the dataframe.
- inspect NaNs, datatypes, and summary statistics

In [179]:
#load the aviation dataset
df = pd.read_csv("data/AviationData.csv", encoding="latin1", low_memory=False)

## Column by column description
1. *Event.Id*: Unique identifier for each accident event.  
2. *Investigation.Type*: Type of investigation (e.g., Accident or Incident).  
3. *Accident.Number*: Official accident reference number.  
4. *Event.Date*: Date when the accident occurred.  
5. *Location*: City or area where the accident happened.  
6. *Country*: Country where the accident occurred.  
7. *Latitude*: Geographic latitude of the accident location.  
8. *Longitude*: Geographic longitude of the accident location.  
9. *Airport.Code*: Code of the nearest airport.  
10. *Airport.Name*: Name of the nearest airport.  
11. *Injury.Severity*: Overall severity classification of the accident.  
12. *Aircraft.damage*: Level of damage to the aircraft.  
13. *Aircraft.Category*: Type or category of the aircraft.  
14. *Registration.Number*: Aircraft registration identifier.  
15. *Make*: Manufacturer of the aircraft.  
16. *Model*: Specific model of the aircraft.  
17. *Amateur.Built*: Indicates whether the aircraft was amateur-built.  
18. *Number.of.Engines*: Number of engines on the aircraft.  
19. *Engine.Type*: Type of engine used by the aircraft.  
20. *FAR.Description*: Regulatory classification of the aircraft.  
21. *Schedule*: Type of flight schedule (e.g., scheduled or non-scheduled).  
22. *Purpose.of.flight*: Purpose of the flight (e.g., personal, business).  
23. *Air.carrier*: Airline or operator of the aircraft.  
24. *Total.Fatal.Injuries*: Number of fatal injuries.  
25. *Total.Serious.Injuries*: Number of serious injuries.  
26. *Total.Minor.Injuries*: Number of minor injuries.  
27. *Total.Uninjured*: Number of people not injured.  
28. *Weather.Condition*: Weather conditions during the accident.  
29. *Broad.phase.of.flight*: Phase of flight when the accident occurred.  
30. *Report.Status*: Status or conclusion of the investigation.  
31. *Publication.Date*: Date when the report was published.

In [180]:
#checking top 5 rows
df.head()

,Event.Id,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,...,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
0,20001218X45444,Accident,SEA87LA080,1948-10-24,"MOOSE CREEK, ID",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,UNK,Cruise,Probable Cause,NaN
1,20001218X45447,Accident,LAX94LA336,1962-07-19,"BRIDGEPORT, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,4.0,0.0,0.0,0.0,UNK,Unknown,Probable Cause,19-09-1996
2,20061025X01555,Accident,NYC07LA005,1974-08-30,"Saltville, VA",United States,36.922223,-81.878056,NaN,NaN,...,Personal,NaN,3.0,NaN,NaN,NaN,IMC,Cruise,Probable Cause,26-02-2007
3,20001218X45448,Accident,LAX96LA321,1977-06-19,"EUREKA, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,IMC,Cruise,Probable Cause,12-09-2000
4,20041105X01764,Accident,CHI79FA064,1979-08-02,"Canton, OH",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,1.0,2.0,NaN,0.0,VMC,Approach,Probable Cause,16-04-1980


In [181]:
#checking bottom 5 rows
df.tail()

,Event.Id,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,...,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
88884,20221227106491,Accident,ERA23LA093,2022-12-26,"Annapolis, MD",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,0.0,1.0,0.0,0.0,NaN,NaN,NaN,29-12-2022
88885,20221227106494,Accident,ERA23LA095,2022-12-26,"Hampton, NH",United States,NaN,NaN,NaN,NaN,...,NaN,NaN,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
88886,20221227106497,Accident,WPR23LA075,2022-12-26,"Payson, AZ",United States,341525N,1112021W,PAN,PAYSON,...,Personal,NaN,0.0,0.0,0.0,1.0,VMC,NaN,NaN,27-12-2022
88887,20221227106498,Accident,WPR23LA076,2022-12-26,"Morgan, UT",United States,NaN,NaN,NaN,NaN,...,Personal,MC CESSNA 210N LLC,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
88888,20221230106513,Accident,ERA23LA097,2022-12-29,"Athens, GA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,0.0,1.0,0.0,1.0,NaN,NaN,NaN,30-12-2022


In [182]:
#checking random 5 rows
df.sample(5)

,Event.Id,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,...,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
70462,20110803X40352,Accident,WPR11LA364,2011-07-27,"Sedona, AZ",United States,345055N,1114719W,SEZ,Sedona Airport,...,Personal,Mary A Nicholson,0.0,0.0,0.0,2.0,VMC,NaN,"The pilot’s improper landing flare, which resu...",25-09-2020
85614,20201117102299,Accident,ANC21LA007,2020-11-16,"Colton, OR",United States,004530N,0122922W,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,IMC,NaN,NaN,09-12-2020
19912,20001213X25211,Accident,ATL88LA124,1988-03-24,"MUSCLE SHOALS, AL",United States,NaN,NaN,MSL,MUSCLE SHOALS,...,Instructional,NaN,0.0,0.0,1.0,0.0,VMC,Cruise,Probable Cause,10-07-1989
67386,20091028X13928,Accident,CEN10LA031,2009-10-27,"Gaylord, MI",United States,045826N,0084402W,GLR,Gaylord Regional,...,Personal,SCREAMIN AVIATION LLC,2.0,0.0,0.0,0.0,VMC,NaN,The pilot's failure to maintain adequate airsp...,25-09-2020
30786,20001211X15117,Accident,CHI92LA206,1992-07-02,"JANESVILLE, WI",United States,NaN,NaN,NaN,NaN,...,Unknown,United Airlines,0.0,1.0,0.0,136.0,IMC,Descent,Probable Cause,09-12-1993


Observation: The dataset is uniform from top to bottom

In [183]:
#Check the data shape
print(f" The dataset has {df.shape[0]} rows and {df.shape[1]} columns")

 The dataset has 88889 rows and 31 columns


In [184]:
#check data type
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50132 non-null  object 
 9   Airport.Name            52704 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87507 non-null  object 
 14  Make                    88826 non-null

Observation:
The dataset has 5 numerical and 26 categorical columns. 
Most columns have complete data, but several columns contain significant missing values. Latitude and Longitude have a large number of missing entries and other columns such as Aircraft.Category and FAR.Description also have many missing values, which may affect analysis and may require cleaning or filtering.


In [185]:
#check for concise statistics on numerical columns
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Number.of.Engines,82805.0,1.146585,0.446510,0.0,1.0,1.0,1.0,8.0
Total.Fatal.Injuries,77488.0,0.647855,5.485960,0.0,0.0,0.0,0.0,349.0
Total.Serious.Injuries,76379.0,0.279881,1.544084,0.0,0.0,0.0,0.0,161.0
Total.Minor.Injuries,76956.0,0.357061,2.235625,0.0,0.0,0.0,0.0,380.0
Total.Uninjured,82977.0,5.325440,27.913634,0.0,0.0,1.0,2.0,699.0


Observation:
- Most aircraft have 1 engine since the median (50%) is 1.
- Fatal, serious, and minor injuries are mostly 0 in many cases, indicating many accidents had no injuries.
- There are extreme injuries, with some accidents having very high counts (fatal injuries up to 349).
- The number of uninjured people varies widely, with some accidents having very large numbers (up to 699).

In [186]:
#check for concise statistics on categorical columns
df.describe(include ='O').T

,count,unique,top,freq
Event.Id,88889,87951,20001212X19172,3
Investigation.Type,88889,2,Accident,85015
Accident.Number,88889,88863,CEN22LA149,2
Event.Date,88889,14782,1984-06-30,25
Location,88837,27758,"ANCHORAGE, AK",434
Country,88663,219,United States,82248
Latitude,34382,25589,332739N,19
Longitude,34373,27154,0112457W,24
Airport.Code,50132,10374,NONE,1488
Airport.Name,52704,24870,Private,240


Observation: 
- Most records are classified as "Accident" under Investigation Type.
- The majority of accidents occurred in the United States.
- Many aircraft are of the make "Cessna", making it the most common manufacturer.
- Most aircraft are not amateur-built.
- The most common engine type is 'Reciprocating'.
- Most flights were for personal purposes.
- Most weather conditions during accidents were "VMC" (good visibility).
- The most common phase of flight during accidents is landing.
- Aircraft damage is mostly classified as "Substantial".
- Many airport codes and registration numbers have missing or "NONE" values.

In [187]:
#check unique values
for coln in df.columns:
    uni_value = df[coln].unique()
    print(f" {coln}\n  {uni_value}\n\n")

 Event.Id
  ['20001218X45444' '20001218X45447' '20061025X01555' ... '20221227106497'
 '20221227106498' '20221230106513']


 Investigation.Type
  ['Accident' 'Incident']


 Accident.Number
  ['SEA87LA080' 'LAX94LA336' 'NYC07LA005' ... 'WPR23LA075' 'WPR23LA076'
 'ERA23LA097']


 Event.Date
  ['1948-10-24' '1962-07-19' '1974-08-30' ... '2022-12-22' '2022-12-26'
 '2022-12-29']


 Location
  ['MOOSE CREEK, ID' 'BRIDGEPORT, CA' 'Saltville, VA' ... 'San Manual, AZ'
 'Auburn Hills, MI' 'Brasnorte, ']


 Country
  ['United States' nan 'GULF OF MEXICO' 'Puerto Rico' 'ATLANTIC OCEAN'
 'HIGH ISLAND' 'Bahamas' 'MISSING' 'Pakistan' 'Angola' 'Germany'
 'Korea, Republic Of' 'Martinique' 'American Samoa' 'PACIFIC OCEAN'
 'Canada' 'Bolivia' 'Mexico' 'Dominica' 'Netherlands Antilles' 'Iceland'
 'Greece' 'Guam' 'Australia' 'CARIBBEAN SEA' 'West Indies' 'Japan'
 'Philippines' 'Venezuela' 'Bermuda' 'San Juan Islands' 'Colombia'
 'El Salvador' 'United Kingdom' 'British Virgin Islands' 'Netherlands'
 'Costa R

Observation:
* Some columns contain a small set of unique values while others have many unique values. 
* Most of the columns, such as latitude, longitude, airport.code, airport.name, among others have null values.

In [188]:
#check column names
df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

Observation
- Column names are consistently formatted using dot notation.
- These column names may need to be standardized using lowercase or replacing dots with underscores for easier handling.

## Data Cleaning

In [189]:
#create a copy of the dataset before doing data cleaning
df1 = df.copy(deep=True)

In [190]:
#standardize column names(removing messy column names)
df1.columns = df1.columns.str.lower().str.strip().str.replace('.', '_', regex=False)
df1.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date'],
      dtype='object')

Observation:
- Column names were converted to lowercase and dots were replaced with underscores.
- This makes the dataset easier to work with and avoids potential syntax issues.

In [191]:
#check missing values in each column
df1.isnull().sum().sort_values(ascending=False)

schedule                  76307
air_carrier               72241
far_description           56866
aircraft_category         56602
longitude                 54516
latitude                  54507
airport_code              38757
airport_name              36185
broad_phase_of_flight     27165
publication_date          13771
total_serious_injuries    12510
total_minor_injuries      11933
total_fatal_injuries      11401
engine_type                7096
report_status              6384
purpose_of_flight          6192
number_of_engines          6084
total_uninjured            5912
weather_condition          4492
aircraft_damage            3194
registration_number        1382
injury_severity            1000
country                     226
amateur_built               102
model                        92
make                         63
location                     52
investigation_type            0
event_date                    0
accident_number               0
event_id                      0
dtype: i

Observation:
- Several columns contain missing values.
- Columns such as latitude, longitude, airport_code,schedule, air_carrier among others have significant missing data.
- Some columns are almost complete with very few missing values.
- Missing data will need to be handled appropriately during cleaning.

In [192]:
#check duplicates
df1.duplicated().sum()

np.int64(0)

There are no duplicate records in the dataset.

### Filtering aircrafts and events

We want to filter the dataset to include aircraft that the client is interested in an analysis of:
- inspect relevant columns
- figure out any reasonable imputations
- filter the dataset

In [193]:
# filter only accident records
df1 = df1[df1['investigation_type'] == 'Accident']
df1["investigation_type"].unique()

array(['Accident'], dtype=object)

Observation: 
- I filtered the dataset to include only accident records and not incidents.
- Incident records were removed as they are not relevant to the analysis.

In [194]:
#remove amateur-built aircraft
df1 = df1[df1['amateur_built'] == 'No']
df1['amateur_built'][:4]

0    No
1    No
2    No
3    No
Name: amateur_built, dtype: object

Observation
- Because the client wants professional aircarft, not homemade one, I removed Amateur-built aircraft from the dataset 
- Only professionally manufactured aircraft were retained for analysis.

In [195]:
#filter by year importance
# convert event_date to datetime
df1['event_date'] = pd.to_datetime(df1['event_date'], errors='coerce')

# filter from 1983 onwards
df1 = df1[df1['event_date'].dt.year >= 1983]
df1['event_date'][:2]

3600   1983-01-01
3601   1983-01-01
Name: event_date, dtype: datetime64[ns]

Observation
- The event_date column was converted to datetime format.
- The dataset was filtered to include records from 1983 onwards as per the client requirements
- Older aircraft were excluded based on the 40-year lifecycle assumption

In [196]:
# filter aircarft category-we require passenger aircraft mainly
#check categories first
df1['aircraft_category'].value_counts()

aircraft_category
Airplane             19937
Helicopter            2934
Glider                 434
Balloon                197
Weight-Shift           139
Powered Parachute       83
Gyrocraft               31
Ultralight              18
Unknown                  9
WSFT                     9
Blimp                    4
Powered-Lift             3
UNK                      2
Rocket                   1
Name: count, dtype: int64

In [197]:
# keep only airplanes
df1 = df1[df1['aircraft_category'] == 'Airplane']
df1["aircraft_category"].unique()

array(['Airplane'], dtype=object)

Observation
- The dataset was filtered to include only airplane category aircraft.
- Other aircraft types were excluded to maintain focus on passenger aircraft.

In [198]:
# drop rows with missing key columns
df1 = df1.dropna(subset=[
    'total_fatal_injuries',
    'total_serious_injuries',
    'aircraft_damage'
])
df1.shape

(16673, 31)

Observation:
- Records with missing values in key columns such as total_fatal_injuries, total_serious_injuries, and aircraft_damage were removed.
- These variables are essential for constructing injury-related metrics and analysis.
- Removing these rows ensures more accurate and reliable results.

In [199]:
# check final dataset shape
df1.shape

(16673, 31)

- The dataset size reduced after filtering.
- The remaining data is more relevant to the client’s objectives.

### Cleaning and constructing Key Measurables

Injuries and robustness to destruction are a key interest point for the client. Clean and impute relevant columns and then create derived fields that best quantifies what the client wishes to track. **Use commenting or markdown to explain any cleaning assumptions as well as any derived columns you create.**

**Construct metric for fatal/serious injuries**

*Hint:* Estimate the total number of passengers on each flight. The likelihood of serious / fatal injury can be estimated as a fraction from this.

In [200]:
# create a 'total number of passengers' column on each flight(both injured and not injured)
df1['total_passengers'] = (
    df1['total_fatal_injuries'] +
    df1['total_serious_injuries'] +
    df1['total_minor_injuries'] +
    df1['total_uninjured']
)
df1['total_passengers'].sum()

np.float64(68343.0)

In [201]:
# create fatal and serious injury rate
df1['fatal_serious_rate'] = (
    (df1['total_fatal_injuries'] + df1['total_serious_injuries']) 
    / df1['total_passengers']
)
(df1['fatal_serious_rate'] == 1).sum()

np.int64(4729)

In [202]:
# handle division by zero or missing values
df1['fatal_serious_rate'] = df1['fatal_serious_rate'].fillna(0)
df1["fatal_serious_rate"].isna().sum()

np.int64(0)

Observation
- A new column total_passengers was created by summing all injury categories.
- A fatal_serious_rate metric was created to measure the severity of accidents.
- Missing or undefined values were replaced with 0 to ensure consistency.

**Aircraft.Damage**
- identify and execute any cleaning tasks
- construct a derived column tracking whether an aircraft was destroyed or not.

In [203]:
#step 1: check unique values in aircraft_damage
df1['aircraft_damage'].value_counts()

aircraft_damage
Substantial    14436
Destroyed       2032
Minor            130
Unknown           75
Name: count, dtype: int64

Observation:
- The aircraft_damage column contains categories such as Substantial, Minor, and Destroyed.
- There are also missing values that need to be handled.
    

In [204]:
#step 2: fill missing values
df1['aircraft_damage'] = df1['aircraft_damage'].fillna('Unknown')

Observation: Missing values in aircraft_damage were replaced with 'Unknown' to retain all records.


In [205]:
#step 3: # clean text
df1['aircraft_damage'] = df1['aircraft_damage'].str.strip().str.lower()

Observation:
Values were standardized by removing whitespace and converting to lowercase for consistency.

In [206]:
#step 4: # create destroyed column
df1['is_destroyed'] = df1['aircraft_damage'].apply(
    lambda x: 1 
    if x == 'destroyed' 
    else 0)
df1['is_destroyed'].value_counts()

is_destroyed
0    14641
1     2032
Name: count, dtype: int64

Observation
- A new column 'is_destroyed' was created to indicate whether an aircraft was destroyed (1) or not (0).
- Most aircraft were not destroyed, with a smaller proportion being completely destroyed.

### Investigate the *Make* column
- Identify cleaning tasks here
- List cleaning tasks clearly in markdown
- Execute the cleaning tasks
- For your analysis, keep Makes with a reasonable number (you can put the threshold at 50 though lower could work as well)

In [207]:
#step 1: # check top values in make column
df1['make'].value_counts().head(20)

make
CESSNA                4727
PIPER                 2747
Cessna                1028
BEECH                  977
Piper                  619
BOEING                 272
MOONEY                 234
Beech                  231
AIR TRACTOR INC        215
CIRRUS DESIGN CORP     206
BELLANCA               156
AERONCA                149
MAULE                  144
LUSCOMBE                93
STINSON                 91
CHAMPION                90
AIR TRACTOR             89
DEHAVILLAND             88
NORTH AMERICAN          76
GRUMMAN                 76
Name: count, dtype: int64

Observation
- The make column contains many unique aircraft manufacturers.
- Some manufacturers appear very frequently like the Cessna while others appear only a few times.

In [208]:
#step 2: # check missing values on the make column
df1['make'].isna().sum()

np.int64(1)

In [209]:
#step 3: # fill missing values
df1['make'] = df1['make'].fillna('Unknown')

Observation:
The make column has only 1 missing value, which was replaced with "Unknown".

In [210]:
#step 4: standardize values
df1['make'] = df1['make'].str.strip().str.lower()

Observation: Values were standardized by converting to lowercase and removing extra spaces.

In [211]:
# step 5: filter ''make' column with at least 50 occurrences

# get counts
make_counts = df1['make'].value_counts()

# keep only makes with >= 50 records
valid_makes = make_counts[make_counts >= 50].index

# filter dataset
df1 = df1[df1['make'].isin(valid_makes)]

In [212]:
make_counts

make
cessna              5755
piper               3366
beech               1208
boeing               311
mooney               298
                    ... 
avions fairey          1
bucker jungmann        1
embraer aircraft       1
mcnulty john s         1
orlican s r o          1
Name: count, Length: 995, dtype: int64

In [213]:
valid_makes

Index(['cessna', 'piper', 'beech', 'boeing', 'mooney', 'air tractor inc',
       'cirrus design corp', 'bellanca', 'maule', 'aeronca', 'air tractor',
       'champion', 'cirrus', 'luscombe', 'stinson', 'grumman',
       'north american', 'dehavilland', 'aero commander', 'aviat aircraft inc',
       'taylorcraft', 'diamond aircraft ind inc', 'socata', 'embraer',
       'grumman acft eng cor-schweizer', 'raytheon aircraft company',
       'rockwell international', 'de havilland', 'american champion aircraft'],
      dtype='object', name='make')

In [214]:
# check remaining makes
df1['make'].value_counts().head()

make
cessna    5755
piper     3366
beech     1208
boeing     311
mooney     298
Name: count, dtype: int64

Observation:
- Aircraft manufacturers with very few records were removed.
- Only manufacturers with at least 50 occurrences were retained.
- This helps reduce noise and focus the analysis on more significant patterns for analysis

### Inspect Model column
- Get rid of any NaNs.
- Inspect the column and counts for each model/make. Are model labels unique to each make?
- If not, create a derived column that is a unique identifier for a given plane type.

In [215]:
#step 1: check missing values
df1['model'].isna().sum()

np.int64(7)

In [216]:
# step 2: fill missing values
df1['model'] = df1['model'].fillna('Unknown')

Observation: The model column has 7 missing values, which were replaced with "Unknown".


In [217]:
df1['model'].head(10)

4171        PA-28-140
11898             152
25379    C99 AIRLINER
25965       PA-28-180
27207            182B
28791           7GCBC
28804           PA-18
30231       DHC-6-200
33535            182A
34275           PA-32
Name: model, dtype: object

In [218]:
# step 3: standardize model values
df1['model'] = df1['model'].str.strip().str.lower()

Observation: I standardized the model column  by converting to lowercase and removing extra spaces in order to remove inconsistensies.

In [219]:
# step 4: check make-model columns and remove any repeated records
df1[['make', 'model']].drop_duplicates().head(20)

,make,model
4171,piper,pa-28-140
11898,cessna,152
25379,beech,c99 airliner
25965,piper,pa-28-180
27207,cessna,182b
28791,bellanca,7gcbc
28804,piper,pa-18
30231,de havilland,dhc-6-200
33535,cessna,182a
34275,piper,pa-32


In [220]:
# step 5: number of models per make
df1.groupby('make')['model'].nunique().sort_values(ascending=False).head()

make
cessna    337
piper     310
beech     196
boeing     94
maule      57
Name: model, dtype: int64

Observation:
- The combination of make and model was inspected to understand their relationship.
- Some manufacturers have multiple models.
- Model names may not be unique across different manufacturers.
- It was observed that models may not be unique across different manufacturers.

In [221]:
# step 6: create unique aircraft type
df1['aircraft_type'] = df1['make'] + '_' + df1['model']

Observation: A new column 'aircraft_type' was created by combining make and model to uniquely identify each aircraft type.

In [222]:
#Step 7: check
df1[['make', 'model', 'aircraft_type']].head()

,make,model,aircraft_type
4171,piper,pa-28-140,piper_pa-28-140
11898,cessna,152,cessna_152
25379,beech,c99 airliner,beech_c99 airliner
25965,piper,pa-28-180,piper_pa-28-180
27207,cessna,182b,cessna_182b


Observation: Each aircraft is now uniquely identified

### Cleaning other columns
- there are other columns containing data that might be related to the outcome of an accident. We list a few here:
- Engine.Type
- Weather.Condition
- Number.of.Engines
- Purpose.of.flight
- Broad.phase.of.flight

Inspect and identify potential cleaning tasks in each of the above columns. Execute those cleaning tasks. 

**Note**: You do not necessarily need to impute or drop NaNs here.

In [223]:
#step 1: check columns
for col in [
    'engine_type',
    'weather_condition',
    'number_of_engines',
    'purpose_of_flight',
    'broad_phase_of_flight']:
    
    print(col)
    print(df1[col].value_counts().head(), "\n")

engine_type
engine_type
Reciprocating    10445
Turbo Prop         720
Turbo Fan          168
Turbo Jet           23
Unknown             13
Name: count, dtype: int64 

weather_condition
weather_condition
VMC    11368
IMC      717
Unk      118
UNK        1
Name: count, dtype: int64 

number_of_engines
number_of_engines
1.0    10880
2.0     1480
4.0       19
3.0        3
Name: count, dtype: int64 

purpose_of_flight
purpose_of_flight
Personal              8186
Instructional         1919
Aerial Application     583
Business               328
Positioning            203
Name: count, dtype: int64 

broad_phase_of_flight
broad_phase_of_flight
Cruise         19
Takeoff        17
Approach       16
Maneuvering     8
Landing         6
Name: count, dtype: int64 



Observation: 
- A quick inspection of selected columns was performed to understand their value distributions.
- Some inconsistencies in formatting (such as capitalization) are observed.

In [224]:
#step 2: Standardize all text to be uniform
cols_to_clean = [
    'engine_type',
    'weather_condition',
    'purpose_of_flight',
    'broad_phase_of_flight']

for col in cols_to_clean:
    df1[col] = df1[col].str.lower().str.strip()

Observation: Text-based columns were standardized by converting all values to lowercase and removing extra spaces. This ensures consistency across categories.

In [225]:
#step 3: converting 'number of engines' to numeric
df1['number_of_engines'] = pd.to_numeric(df1['number_of_engines'], errors='coerce')

Observation: The number_of_engines column was converted to numeric format to ensure it can be used for analysis.

In [226]:
#step 4: Check cleaned values
df1['engine_type'].value_counts().head()

engine_type
reciprocating    10445
turbo prop         720
turbo fan          168
turbo jet           23
unknown             13
Name: count, dtype: int64

In [227]:
df1['weather_condition'].value_counts().head()

weather_condition
vmc    11368
imc      717
unk      119
Name: count, dtype: int64

Observation:
- After cleaning, categorical values are now consistently formatted.
- Duplicate categories caused by inconsistent casing or spacing have been resolved.
- The data is now more reliable for grouping and analysis.

In [228]:
#step 5: Ensure all unknowns are uniform
df1 = df1.replace(['Unknown', 'unknown'], 'unknown')

Observation: Variations of the "unknown" category were standardized to a single and uniform consistent label.

Summary of cleaning
- Key categorical columns were cleaned and standardized.
- Formatting inconsistencies were resolved.
- Numerical columns were converted to appropriate data types.
- The dataset is now more structured and ready for analysis.

### Column Removal
- inspect the dataframe and drop any columns that have too many NaNs

In [229]:
#step 1: check for missing values
df1.isna().sum().sort_values(ascending=False)

broad_phase_of_flight     13429
schedule                  12607
air_carrier                6385
airport_code               4241
airport_name               4172
report_status              2717
engine_type                2131
purpose_of_flight          1407
weather_condition          1302
number_of_engines          1124
longitude                   694
latitude                    694
publication_date            429
injury_severity             151
far_description             136
total_passengers             37
total_uninjured              35
total_minor_injuries         32
registration_number          15
total_fatal_injuries          0
fatal_serious_rate            0
total_serious_injuries        0
is_destroyed                  0
event_id                      0
investigation_type            0
amateur_built                 0
model                         0
make                          0
aircraft_category             0
aircraft_damage               0
country                       0
location

Observation:
- Some columns contain a high number of missing values.

Observation:
* Based on the dataset; airport_code, airport_name, latitude and longitude have high number of missing values and may not be useful for analysis.


In [230]:
#Step 2: Drop columns 
columns_to_drop = [
    'airport_code',
    'airport_name',
    'latitude',
    'longitude'
]

df1 = df1.drop(columns=columns_to_drop)

In [231]:
#step 3: Check
df1.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date', 'total_passengers', 'fatal_serious_rate',
       'is_destroyed', 'aircraft_type'],
      dtype='object')

Observation:
- Columns airport_code, airport_name, latitude, and longitude were removed due to a high proportion of missing values.
- These features were not essential to the core analysis of injury severity and aircraft damage.

### Save DataFrame to csv
- its generally useful to save data to file/server after its in a sufficiently cleaned or intermediate state
- the data can then be loaded directly in another notebook for further analysis
- this helps keep your notebooks and workflow readable, clean and modularized

In [232]:
#saving the clean dataset
df1.to_csv('clean_aviation_data.csv', index=False)